### PR TITLE
use a structured suggestion for char-lit-as-u8

### PR DESCRIPTION
**Pull request by euclio**
changelog: use a structured suggestion for char-lit-as-u8

### CODE REVIEW

File: tests/ui/char_lit_as_u8.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// run-rustfix
+
 #![warn(clippy::char_lit_as_u8)]
 #![allow(unused_variables)]
 fn main() {
```
Comment (phansch) [inline tests/ui/char_lit_as_u8.rs]: I'd feel a bit safer about the suggestions with a couple more tests here:

```rust
// These should all trigger the lint
let _ = '❤' as u8;
let _ = '\n' as u8;
let _ = '\0' as u8;
let _ = '\x01' as u8;
```

**Comment by euclio** [comment]
Just pushed some more tests. I also tweaked the wording a bit more.

File: clippy_lints/src/types.rs [hunk 2]
```diff
@@ -1473,29 +1472,39 @@ declare_clippy_lint! {
...
-                            snippet(cx, e.span, "'x'")
-                        );
-                        span_help_and_lint(cx, CHAR_LIT_AS_U8, expr.span, msg, &help);
-                    }
-                }
+        if_chain! {
+            if let ExprKind::Cast(e, _) = &expr.node;
+            if let ExprKind::Lit(l) = &e.node;
+            if let LitKind::Char(c) = l.node;
+            if ty::Uint(UintTy::U8) == cx.tables.expr_ty(expr).sty && !expr.span.from_expansion();
```
Comment (flip1995) [inline clippy_lints/src/types.rs]: Can you move the `from_expansion()` check up to the beginning of the `if_chain!` so the destructuring can be skipped for expanded code?

Comment (euclio) [reply clippy_lints/src/types.rs]: Fixed.

File: clippy_lints/src/types.rs [hunk 3]
```diff
@@ -1473,29 +1472,40 @@ declare_clippy_lint! {
     /// ```
     pub CHAR_LIT_AS_U8,
     complexity,
-    "casting a character literal to u8"
+    "casting a character literal to u8 truncates"
```
Comment (flip1995) [inline clippy_lints/src/types.rs]: You have to run `util/dev update_lints` since you changed a lint description.

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Pull request merged**